conductor-modules/viewpager2/src/main/java/com/bluelinelabs/conductor/viewpager/RouterStateAdapter.kt

```diff
@@ -0,0 +1,221 @@
+
+    attachRouter(holder, position)
+  }
+
+  override fun onViewAttachedToWindow(holder: RouterViewHolder) {
+    super.onViewAttachedToWindow(holder)
+
+    if (!holder.attached) {
+      attachRouter(holder, holder.currentItemPosition)
+    }
+  }
+
+  override fun onViewDetachedFromWindow(holder: RouterViewHolder) {
+    super.onViewDetachedFromWindow(holder)
+
+    detachRouter(holder)
+  }
+
+  override fun onViewRecycled(holder: RouterViewHolder) {
+    super.onViewRecycled(holder)
+
+    detachRouter(holder)
+
+    holder.currentRouter?.let { router ->
+      host.removeChildRouter(router)
+      holder.currentRouter = null
+    }
+  }
+
+  override fun onFailedToRecycleView(holder: RouterViewHolder): Boolean {
+    return true
+  }
+
+  override fun getItemId(position: Int): Long {
+    return position.toLong()
+  }
+
+  override fun setHasStableIds(hasStableIds: Boolean) {
+    throw UnsupportedOperationException("Stable Ids are required for the adapter to function properly")
+  }
+
+  override fun saveState(): Parcelable {
+    return SavedState(
+      savedPagesKeys = (0 until savedPages.size()).map { savedPages.keyAt(it) },
+      savedPagesValues = (0 until savedPages.size()).map { savedPages.valueAt(it) },
+      savedPageHistory = savedPageHistory,
+      maxPagesToStateSave = maxPagesToStateSave
+    )
+  }
+
+  override fun restoreState(state: Parcelable) {
+    if (state !is SavedState) return
+
+    savedPages = LongSparseArray()
+    state.savedPagesKeys.indices.forEach { index ->
+      savedPages.put(state.savedPagesKeys[index], state.savedPagesValues[index])
+    }
+
+    savedPageHistory = state.savedPageHistory.toMutableList()
+    maxPagesToStateSave = state.maxPagesToStateSave
+  }
+
+  private fun attachRouter(holder: RouterViewHolder, position: Int) {
+    val itemId = getItemId(position)
+    val router = host.getChildRouter(holder.container, "$itemId")
+    holder.currentRouter = router
+    holder.currentItemId = itemId
+
+    if (!router.hasRootController()) {
+      val routerSavedState = savedPages[position.toLong()]
+      if (routerSavedState != null) {
+        router.restoreInstanceState(routerSavedState)
+        savedPages.remove(itemId)
+        savedPageHistory.remove(itemId)
+      }
+    }
+
+    router.rebindIfNeeded()
+    configureRouter(router, position)
+
+    if (position != currentPrimaryRouterPosition) {
+      for (transaction in router.backstack) {
+        transaction.controller.setOptionsMenuHidden(true)
+      }
+    }
+
+    visibleRouters.put(position, router)
+
+    holder.attached = true
+  }
+
+  private fun detachRouter(holder: RouterViewHolder) {
+    holder.currentRouter?.let { router ->
+      router.prepareForHostDetach()
+
+      val savedState = Bundle()
+      router.saveInstanceState(savedState)
+      savedPages.put(holder.currentItemId, savedState)
+
+      savedPageHistory.remove(holder.currentItemId)
+      savedPageHistory.add(holder.currentItemId)
```

Comment:
@EricKuck remove then add? should it just be `savedPageHistory.add(holder.currentItemId)`